README.md
![overview](https://github.com/daaaanil81/Diploma/blob/master/public/images/logo2.png)
# Diploma
My Diploma work
In this repositorie, I keep my diploma work, because I don`t want his to lose.
In camera_file directory:
  1. export IP='your ip address for server'
  2. make start_http_server // starting server with https protocol
  3. make compile // compiling source files
  4. make clean // clear executable file
# For start program
  1. cd Diploma
  2. make compile
  3. node server.js IP
  4. ./tcp_server/ws-server IP
# Programe will compile with openssl with different parameters -lssl -lcrypto
  1. cd /usr/local/include/
  2. ln -s ../opt/openssl/include/openssl .
  3. export LIBRARY_PATH=$LIBRARY_PATH:/usr/local/opt/openssl/lib/
# For create certificate:
  openssl req -new -newkey rsa:4096 -days 36500 -nodes -x509 -keyout "localhost-100y.key" -out "localhost-100y.cert
# For install libwebsockets
  1. git clone -b v3.2-stable https://github.com/warmcat/libwebsockets.git
  2. cd libwebsockets/ 
  3. mkdir build 
  4. cd build
  5. cmake ..
  6. make
  7. sudo make install    


Docs/history.txt
    1  cd libwebsockets/
    2  ls
    3  cd build/
    4  ls
    5  sudo make install
    6  sudo passwd root
    7  cd Diploma/
    8  ls
    9  cd ..
   10  git clone 
   11  git clone https://github.com/warmcat/libwebsockets.git
   12  cd libwebsockets/
   13  ls
   14  ./config
   15  mkdir build
   16  cd build/
   17  cmake ..
   18  sudo apt-get install cmake
   19  cmake ..
   20  make && sudo make install
   21  cd ..
   22  libwebsockets --version
   23  libwebsockets -v
   24  ./libwebsockets -v
   25  ./libwebsockets --version
   26  cd Diploma/
   27  cd camera_file/
   28  ls
   29  sudo apt-get install nodejs
   30  ls
   31  node ./js/http_server.js 
   32  node ./js/http_server.js 10.168.166.132
   33  ./rtpengine --interface=10.168.166.132 --listen-ng=127.0.0.1:2223 --port-min=12301 --port-max=14999 -f -E
   34  cd rtpengine/daemon/
   35  ./rtpengine --interface=10.168.166.132 --listen-ng=127.0.0.1:2223 --port-min=12301 --port-max=14999 -f -E
   36  ls
   37  ifconfig 
   38  ls
   39  uname -a
   40  pwd
   41  sudo apt-get install git
   42  su
   43  ls -l
   44  git clone https://github.com/daaaanil81/Diploma.git
   45  ls
   46  cd Diploma/
   47  ls
   48  cd ..
   49  ls
   50  git https://github.com/sipwise/rtpengine.git
   51  git clone https://github.com/sipwise/rtpengine.git
   52  ls
   53  cd rtpengine/
   54  ls
   55  vim Makefile 
   56  cd daemon/
   57  ls
   58  vim Makefile 
   59  make 
   60  sudo apt-get install make
   61  sudo apt-get install make-guile
   62  sudo apt-get update
   63  sudo apt-get upgrade
   64  make 
   65  make -j4
   66  ls
   67  c d..
   68  cd ..
   69  make -j4
   70  sudo apt-get install pkg-config
   71  make clean
   72  sudo apt-get install libssl-dev
   73  make clean
   74  make -j4
   75  sudo apt-get install libglib2.0-dev
   76  ді
   77  make -j4
   78  sudo apt-get install libpcap0.8-dev
   79  make -j4
   80  sudo apt-get install libhiredis-dev
   81  make -j4
   82  cd ..
   83  git clone 
   84  git clone https://github.com/syntheticpp/CMakeLua.git
   85  cd Cm
   86  cd CMakeLua/
   87  ./bootstrap 
   88  sudo apt-get install libxmlrpc-core-c3-dev
   89  cd ..
   90  rm -rf CMakeLua/
   91  cd rtpengine/
   92  make -j4
   93  sudo apt-get install libevent-dev
   94  make -j4
   95  sudo apt-get install libjson-glib-dev
   96  make -j4
   97  sudo apt-get install libspandsp-dev
   98  make -j4
   99  sudo apt-get install libcurl3 -y
  100  make -j4
  101  sudo apt-get install libcurl4 libcurl4-openssl-dev -y
  102  make -j4
  103  make clean
  104  sudo apt-get install libavutil-dev
  105  make clean
  106  make clear
  107  cd daemon/
  108  make 
  109  make clear
  110  sudo apt-get install libxmlrpc-core-c3-dev
  111  make clear
  112  make -j3
  113  make -j4
  114  cd ..
  115  rm -rf rtpengine/
  116  git clone -b mr8.0.1 --single-branch https://github.com/sipwise/rtpengine.git
  117  cd rtpengine/
  118  git branch
  119  make -j4
  120  sudo apt-get install libswresample-dev
  121  l
  122  ls
  123  make -j4
  124  make cleanr
  125  make clear
  126  sudo apt-get install ffmpeg
  127  make -j4
  128  sudo apt-get install libavcodec 
  129  make -j4
  130  sudo apt-get install libavcodec 
  131  sudo apt-get install -y libavcodec-extra57
  132  make -j4
  133  sudo add-apt-repository ppa:jonathonf/ffmpeg-4
  134  sudo apt-get install ffmpeg
  135  make -j4
  136  sudo apt-get install libavcodec-dev
  137  make -j4
  138  sudo apt-get install libavformat53 libavformat-dev libavcodec53
  139  sudo apt install make gcc pkg-config meson
  140  sudo apt-get install ibavformat-dev libavutil-dev
  141  sudo apt-get install libavformat-dev libavutil-dev
  142  make -j4
  143  sudo apt install libiptc0 libiptc-dev sudo apt-get install mysql-client
  144  make -j4
  145  sudo apt install mqsql
  146  sudo apt install mysql
  147  sudo apt-get install mysql-client
  148  sudo apt-get install mysql-server
  149  make -j4
  150  pt-get install libmysqlclient-dev 
  151  sudo apt-get install libmysqlclient-dev 
  152  make -j4
  153  make clear
  154  make clean
  155  sudo apt-get install libavfilter-devsudo apt-get install libavfilter-dev
  156  sudo apt-get install libavfilter-dev
  157  make -j4
  158  ls
  159  cd debian/
  160  ls
  161  cd ..
  162  cd daemon/
  163  ls
  164  make clean 
  165  ls
  166  cd ..
  167  make clear
  168  make clean
  169  ls
  170  cd daemon/
  171  ls
  172  make -j4
  173  ls
  174  make
  175  ls
  176  sudo apt-get install perl
  177  sudo apt-get install perl5
  178  sudo apt-get update
  179  sudo apt-get upgrad
  180  sudo apt-get upgrade
  181  ls
  182  make 
  183  sudo apt-get install -y gperf
  184  make 
  185  make clear
  186  make clean
  187  make 
  188  ls
  189  cd ..
  190  ls
  191  cd Diploma/p
  192  ls
  193  cd Diploma/
  194  ls
  195  cd camera_file/
  196  ls
  197  make compile
  198  ./tcp_server/ws-server 10.168.166.132
  199  cd /usr/local/include/
  200  ln -s ../opt/openssl/include/openssl .
  201  sudo ln -s ../opt/openssl/include/openssl .
  202  export LIBRARY_PATH=$LIBRARY_PATH:/usr/local/opt/openssl/lib/
  203  cd ~/Diploma/camera_file/
  204  ./tcp_server/ws-server 10.168.166.132
  205  echo LD_LIBRARY_PATH
  206  echo $LD_LIBRARY_PATH
  207  ls
  208  export $LD_LIBRARY_PATH+=/usr/local/lib64
  209  export LD_LIBRARY_PATH+=/usr/local/lib64
  210  echo $LD_LIBRARY_PATH
  211  ./tcp_server/ws-server 10.168.166.132
  212  grep -R so /
  213  grep -R libwebsockets  /
  214  grep -R libwebsockets.so  /
  215  ./tcp_server/ws-server 10.168.166.132
  216  ls
  217  cd ..
  218  l
  219  scd libwebsockets/
  220  cd ..
  221  cd 
  222  rm -rf libwebsockets/
  223  git clone -b v3.2-stable https://github.com/warmcat/libwebsockets.git
  224  git branch
  225  cd libwebsockets/
  226  git branch
  227  mkdir build 
  228  cd build
  229  cd ../..
  230  git clone https://github.com/microsoft/vcpkg.git
  231  cd libwebsockets/

tcp_server/rtp.c
#include "rtp.h"

int crypto_encrypt_rtp(struct crypto_context *c, struct str_key *payload, uint32_t ssrc, uint64_t index)
{
	return c->params.crypto_suite->encrypt_rtp(c, payload, ssrc, index);
}
int crypto_hash_rtp(struct crypto_context *c, unsigned char *payload, struct str_key *mes_all, uint64_t index)
{
	return c->params.crypto_suite->hash_rtp(c, payload, mes_all, index);
}
int crypto_init_session_key_rtp(struct crypto_context *c)
{
	return c->params.crypto_suite->session_key_init(c);
}
void rtp_init(struct pthread_arguments *p_a)
{
	p_a->sequnce_new = 0;
	p_a->sequnce_origin = 0;
	p_a->index_rtcp = 0;
	p_a->index = 0;
}

unsigned int rtp_sps_parse(unsigned char *rtp, unsigned char *sps, unsigned int sequnce, struct pthread_arguments *p_a)
{
	unsigned int seq_num = sequnce; // sequnce before I-Frame
	unsigned short index = 0;
	unsigned char h = rtp[2];
	unsigned char l = rtp[3];
	for (int i = 0; i < 2; i++)
	{
		sps[index++] = rtp[i];
	}
	sps[index++] = seq_num >> 8;
	sps[index++] = seq_num;
	for (int i = 0; i < 8; i++)
	{
		sps[index++] = rtp[i + 4];
	}
	sps[index++] = 56; // 0x38
	sps[index++] = (p_a->size_sps) & 0xFF >> 8;
	sps[index++] = (p_a->size_sps) & 0xFF;
	//sps[index++] = 39; // 0x27 Nal header byte with sps payload
	for (int i = 0; i < p_a->size_sps; i++)
	{
		sps[index++] = p_a->sps[i];
	}
	sps[index++] = (p_a->size_pps) & 0xFF >> 8;
	sps[index++] = (p_a->size_pps) & 0xFF;
	//sps[index++] = 40; // 0x28 Nal header byte with pps payload
	for (int i = 0; i < p_a->size_pps; i++)
	{
		sps[index++] = p_a->pps[i];
	}
	return index;
}
/** rtp_payload fill payload and sps
 * struct rtp_header save data about rtp header  
 * 
*/
unsigned int rtp_payload(struct rtp_header *rtp, struct str_key *payload, uint16_t *sequnce_origin, uint16_t *sequnce_new,
						 unsigned char *all_mess, struct str_key *rtp_sps, struct pthread_arguments *p_a, int l)
{
	unsigned char v = (all_mess[0] & 0xC0) >> 6;
	unsigned char p = (all_mess[0] & 0x20) >> 5;
	unsigned char x = (all_mess[0] & 0x10) >> 4;
	unsigned char cc = (all_mess[0] & 0x0F);
	unsigned char m = (all_mess[1] & 0x80) >> 7;
	unsigned char payload_type = all_mess[1] & 0x7F;
	unsigned int seq_num = (all_mess[2] & 0xFF) << 8 | (all_mess[3] & 0xFF); // 1054i // 1055 // 1056 // 1057i // 1058
	uint32_t timestamp = (all_mess[4] & 0xFF) << 24 | (all_mess[5] & 0xFF) << 16 | (all_mess[6] & 0xFF) << 8 | (all_mess[7] & 0xFF);
	uint32_t ssrc = (all_mess[8] & 0xFF) << 24 | (all_mess[9] & 0xFF) << 16 | (all_mess[10] & 0xFF) << 8 | (all_mess[11] & 0xFF);
	unsigned int sq_copy = *sequnce_new;
	unsigned short indicator_type = all_mess[12] & 0x1F;
	unsigned short header_type = all_mess[13] & 0x1F;
	unsigned short header_start = (all_mess[13] & 0x80) >> 7;
	all_mess[1] = (all_mess[1] & 0x80) | 102; /** Change payload in rtp header */
	rtp->all_mark[0] = all_mess[0];
	rtp->all_mark[1] = all_mess[1];
	rtp->seq_num = seq_num;
	rtp->ssrc = ssrc;
	rtp->timestamp = timestamp;
	if (FLAG_TESTING)
	{
		printf("V = %u\n", v);
		printf("P = %u\n", p);
		printf("X = %u\n", x);
		printf("CC = %u\n", cc);
		/* 1 byte */
		printf("M = %u\n", m);
		printf("Payload tupe = %u\n", payload_type);
		/* 2-3 bytes */
		printf("Sequence number = %u\n", seq_num);
		/* 4-7 bytes */
		printf("Timestamp = %u\n", timestamp);
		/* 8-11 bytes */
		printf("SSRC = %u\n", ssrc);
	}
	payload->str = all_mess + 12;

	//printf("Old sequnce: %d\n", *sequnce_new);

	if (*sequnce_origin == 0)
	{
		//printf("Start rtp_sequnce\n");
		*sequnce_origin = rtp->seq_num;
		sq_copy = *sequnce_new = 1;
	}
	if (indicator_type == 28 && header_type == 5 && header_start == 1)
	{
		//printf("I-frame\n");
		rtp_sps->len = rtp_sps_parse(all_mess, rtp_sps->str, sq_copy, p_a);
		sq_copy++;
	}
	all_mess[2] = (sq_copy & 0xFF00) >> 8;
	all_mess[3] = sq_copy & 0xFF;
	//printf("New sequnce: %d\n", *sequnce_new);
	return l - 12;
}
int check_session_keys_rtp(struct crypto_context *c)
{
	if (c->have_session_key)
		return 0;

	struct str_key s;

	s.str = (unsigned char *)c->session_key;
	s.len = c->params.crypto_suite->session_key_len;
	if (crypto_gen_session_key(c, &s, 0x00, 6))
		goto error;
	s.str = (unsigned char *)c->session_auth_key;
	s.len = c->params.crypto_suite->srtp_auth_key_len;
	if (crypto_gen_session_key(c, &s, 0x01, 6))
		goto error;
	s.str = (unsigned char *)c->session_salt;
	s.len = c->params.crypto_suite->session_salt_len;
	if (crypto_gen_session_key(c, &s, 0x02, 6))
		goto error;
	c->have_session_key = 1;
	crypto_init_session_key_rtp(c);

	return 0;

error:
	printf("Error in check_session_keys\n");
	return -1;
}
static u_int64_t packet_index(uint16_t seq_num, uint64_t srtp_index)
{
	u_int16_t seq = seq_num;
	if (srtp_index == 0)
		srtp_index = seq;
	/* rfc 3711 appendix A, modified, and sections 3.3 and 3.3.1 */
	u_int16_t s_l = (srtp_index & 0x00000000ffffULL);
	u_int32_t roc = (srtp_index & 0xffffffff0000ULL) >> 16;
	u_int32_t v = 0;

	if (s_l < 0x8000)
	{
		if (((seq - s_l) > 0x8000) && roc > 0)
			v = (roc - 1) % 0x10000;
		else
			v = roc;
	}
	else
	{
		if ((s_l - 0x8000) > seq)
			v = (roc + 1) % 0x10000;
		else
			v = roc;
	}
	srtp_index = (u_int64_t)(((v << 16) | seq) & 0xffffffffffffULL);
	return srtp_index;
}
/**
 * rtp_to_srtp --> do all work with rtp packets
 * p_a -->struct with all information
 * rtp --> array data with rtp packet
 * l --> size 0f array data
*/
int rtp_to_srtp(struct pthread_arguments *p_a, unsigned char *rtp, unsigned char *rtp_sps, int *l)
{
	uint64_t index = p_a->index;
	struct str_key payload; /** Payload for save payload from camera and crypt his and send to browser */
	struct str_key sps_mes;
	struct str_key sps_payload;
	struct rtp_header rtp_h; /** Rtp_h struct for save rtp header */
	struct str_key mes_all;  /** mes_all save all rtp message */
	int socket_rtp;
	sps_mes.str = rtp_sps; /** Struct with all sps data*/
	sps_mes.len = 0;

	payload.len = rtp_payload(&rtp_h, &payload, &p_a->sequnce_origin, &p_a->sequnce_new, rtp, &sps_mes, p_a, *l); /** Fill rtp_h and payload structs from array bytes from camera */

	if (sps_mes.len != 0)
	{
		//printf("Ok\n");
		sps_payload.str = sps_mes.str + 12;
		sps_payload.len = sps_mes.len - 12;
		check_session_keys_rtp(&p_a->crypto);
		index = p_a->index = packet_index(p_a->sequnce_new, index); /// 1 ...
		crypto_encrypt_rtp(&p_a->crypto, &sps_payload, rtp_h.ssrc, index);
		crypto_hash_rtp(&p_a->crypto, (unsigned char *)(sps_mes.str + sps_mes.len), &sps_mes, index);
		sps_mes.len += 10;
		p_a->sequnce_new += 1; /// 2 ...
		if(p_a->sequnce_new == 0)
			printf("Error ------ Error Sequnce = 0\n");
	}

	mes_all.str = rtp;
	mes_all.len = *l;


	check_session_keys_rtp(&p_a->crypto);
	index = p_a->index = packet_index(p_a->sequnce_new, index);

	crypto_encrypt_rtp(&p_a->crypto, &payload, rtp_h.ssrc, index);
	/** Function Hash all mess and add    
	 * payload->str + payload->len --- place where will save 10 bytes
	 * mes_all --- all message
	*/
	crypto_hash_rtp(&p_a->crypto, (unsigned char *)(mes_all.str + mes_all.len), &mes_all, index);
	payload.len += p_a->crypto.params.crypto_suite->srtp_auth_tag;
	mes_all.len += p_a->crypto.params.crypto_suite->srtp_auth_tag;
	p_a->sequnce_new += 1;
	if(p_a->sequnce_new == 0)
		printf("Error ------ Error Sequnce = 0\n");
	*l = mes_all.len;
	return sps_mes.len;
}
